From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Muthukumar <mk@verveflux.com>
Date: Sun, 8 Feb 2026 07:34:11 +0530
Subject: feat: add override hooks for 10 APIs (Batch 5)

Element: clientWidth, clientHeight, scrollWidth, scrollHeight
HTMLElement: offsetWidth, offsetHeight, offsetLeft, offsetTop
Screen: width, height

diff --git a/third_party/blink/renderer/core/display_lock/display_lock_utilities_test.cc b/third_party/blink/renderer/core/display_lock/display_lock_utilities_test.cc
index 9d7712dc9fb61a7d47e0ca576edfa020befaa28c..997346e4391b5fc9c8a5a2c0ede3799024f8dbbe 100644
--- a/third_party/blink/renderer/core/display_lock/display_lock_utilities_test.cc
+++ b/third_party/blink/renderer/core/display_lock/display_lock_utilities_test.cc
@@ -346,7 +346,7 @@ TEST_F(DisplayLockUtilitiesTest, ContainerQueryCrash) {
   ASSERT_TRUE(child);
 
   // Should not fail DCHECKs or crash.
-  child->offsetTopForBinding();
+  child->offsetTopForBinding(nullptr);
 }
 
 }  // namespace blink
diff --git a/third_party/blink/renderer/core/dom/element.cc b/third_party/blink/renderer/core/dom/element.cc
index ee83e8955f5b6dde7519552681cdd9b707dc0bd8..d6c252f9320afeab8bf079235a93ea68ad13e1db 100644
--- a/third_party/blink/renderer/core/dom/element.cc
+++ b/third_party/blink/renderer/core/dom/element.cc
@@ -2493,6 +2493,50 @@ int Element::clientHeight() {
   return result;
 }
 
+namespace {
+int MaybeIntOverride(ScriptState* script_state,
+                     int value,
+                     const char* interface_key,
+                     const char* override_name) {
+  if (!script_state || !script_state->ContextIsValid())
+    return value;
+
+  v8::Isolate* isolate = script_state->GetIsolate();
+  v8::HandleScope handle_scope(isolate);
+  ScriptState::Scope scope(script_state);
+
+  v8::Local<v8::Function> override_fn =
+      GetCustomOverrideFunction(script_state, interface_key, override_name);
+  if (override_fn.IsEmpty())
+    return value;
+
+  v8::Local<v8::Context> context = script_state->GetContext();
+  v8::Local<v8::Value> argv[] = {
+      v8::Number::New(isolate, static_cast<double>(value))};
+  v8::TryCatch try_catch(isolate);
+  v8::Local<v8::Value> override_result;
+  if (!V8ScriptRunner::CallFunction(
+           override_fn, ExecutionContext::From(script_state),
+           context->Global(), 1, argv, isolate)
+           .ToLocal(&override_result))
+    return value;
+
+  if (override_result->IsNumber())
+    return static_cast<int>(override_result.As<v8::Number>()->Value());
+  return value;
+}
+}  // namespace
+
+int Element::clientWidthForBinding(ScriptState* script_state) {
+  return MaybeIntOverride(script_state, clientWidth(), "element",
+                          "clientWidthOverride");
+}
+
+int Element::clientHeightForBinding(ScriptState* script_state) {
+  return MaybeIntOverride(script_state, clientHeight(), "element",
+                          "clientHeightOverride");
+}
+
 double Element::currentCSSZoom() {
   GetDocument().UpdateStyleAndLayoutTreeForElement(
       this, DocumentUpdateReason::kComputedStyle);
@@ -2755,6 +2799,16 @@ int Element::scrollHeight() {
   return 0;
 }
 
+int Element::scrollWidthForBinding(ScriptState* script_state) {
+  return MaybeIntOverride(script_state, scrollWidth(), "element",
+                          "scrollWidthOverride");
+}
+
+int Element::scrollHeightForBinding(ScriptState* script_state) {
+  return MaybeIntOverride(script_state, scrollHeight(), "element",
+                          "scrollHeightOverride");
+}
+
 ScriptPromise<IDLUndefined> Element::scrollBy(ScriptState* script_state,
                                               double x,
                                               double y) {
diff --git a/third_party/blink/renderer/core/dom/element.h b/third_party/blink/renderer/core/dom/element.h
index d1f2cae405164cac338c57b7077bb3d680b19307..229f6c5bd82266ff8a87f75899ba288444915c67 100644
--- a/third_party/blink/renderer/core/dom/element.h
+++ b/third_party/blink/renderer/core/dom/element.h
@@ -535,6 +535,8 @@ class CORE_EXPORT Element : public ContainerNode, public Animatable {
   int ClientTopNoLayout() const;
   int clientWidth();
   int clientHeight();
+  int clientWidthForBinding(ScriptState*);
+  int clientHeightForBinding(ScriptState*);
   double currentCSSZoom();
   double scrollLeft();
   double scrollTop();
@@ -542,6 +544,8 @@ class CORE_EXPORT Element : public ContainerNode, public Animatable {
   void setScrollTop(double);
   virtual int scrollWidth();
   virtual int scrollHeight();
+  int scrollWidthForBinding(ScriptState*);
+  int scrollHeightForBinding(ScriptState*);
 
   ScriptPromise<IDLUndefined> scrollIntoView(
       ScriptState* script_state,
diff --git a/third_party/blink/renderer/core/dom/element.idl b/third_party/blink/renderer/core/dom/element.idl
index b9f8ad7e466baac7ab82a45d0b45aff9b5d06a09..e22d692d94bd99bf2135f859d1dd37b48371ed7c 100644
--- a/third_party/blink/renderer/core/dom/element.idl
+++ b/third_party/blink/renderer/core/dom/element.idl
@@ -145,12 +145,12 @@ dictionary SetHTMLUnsafeOptions {
     [CallWith=ScriptState] Promise<undefined> scrollBy(unrestricted double x, unrestricted double y);
     attribute unrestricted double scrollTop;
     attribute unrestricted double scrollLeft;
-    readonly attribute long scrollWidth;
-    readonly attribute long scrollHeight;
+    [CallWith=ScriptState, ImplementedAs=scrollWidthForBinding] readonly attribute long scrollWidth;
+    [CallWith=ScriptState, ImplementedAs=scrollHeightForBinding] readonly attribute long scrollHeight;
     readonly attribute long clientTop;
     readonly attribute long clientLeft;
-    readonly attribute long clientWidth;
-    readonly attribute long clientHeight;
+    [CallWith=ScriptState, ImplementedAs=clientWidthForBinding] readonly attribute long clientWidth;
+    [CallWith=ScriptState, ImplementedAs=clientHeightForBinding] readonly attribute long clientHeight;
     [RuntimeEnabled=StandardizedBrowserZoom] readonly attribute double currentCSSZoom;
 
     // Used by both Anchor Positioning and Popover
diff --git a/third_party/blink/renderer/core/frame/screen.cc b/third_party/blink/renderer/core/frame/screen.cc
index d81e607d817edb352b07548cac3c595e966c7edf..90516630f4cb0e758f1e0efce222373fb260189d 100644
--- a/third_party/blink/renderer/core/frame/screen.cc
+++ b/third_party/blink/renderer/core/frame/screen.cc
@@ -35,6 +35,8 @@
 #include "third_party/blink/renderer/core/frame/local_frame.h"
 #include "third_party/blink/renderer/core/frame/settings.h"
 #include "third_party/blink/renderer/core/page/chrome_client.h"
+#include "third_party/blink/renderer/bindings/core/v8/v8_script_runner.h"
+#include "third_party/blink/renderer/platform/bindings/custom_override_helpers.h"
 #include "ui/display/screen_info.h"
 #include "ui/display/screen_infos.h"
 
@@ -90,6 +92,48 @@ int Screen::width() const {
   return GetRect(/*available=*/false).width();
 }
 
+namespace {
+int MaybeIntOverride(ScriptState* script_state,
+                     int value,
+                     const char* interface_key,
+                     const char* override_name) {
+  if (!script_state || !script_state->ContextIsValid())
+    return value;
+
+  v8::Isolate* isolate = script_state->GetIsolate();
+  v8::HandleScope handle_scope(isolate);
+  ScriptState::Scope scope(script_state);
+
+  v8::Local<v8::Function> override_fn =
+      GetCustomOverrideFunction(script_state, interface_key, override_name);
+  if (override_fn.IsEmpty())
+    return value;
+
+  v8::Local<v8::Context> context = script_state->GetContext();
+  v8::Local<v8::Value> argv[] = {
+      v8::Number::New(isolate, static_cast<double>(value))};
+  v8::TryCatch try_catch(isolate);
+  v8::Local<v8::Value> override_result;
+  if (!V8ScriptRunner::CallFunction(
+           override_fn, ExecutionContext::From(script_state),
+           context->Global(), 1, argv, isolate)
+           .ToLocal(&override_result))
+    return value;
+
+  if (override_result->IsNumber())
+    return static_cast<int>(override_result.As<v8::Number>()->Value());
+  return value;
+}
+}  // namespace
+
+int Screen::heightForBinding(ScriptState* script_state) const {
+  return MaybeIntOverride(script_state, height(), "screen", "heightOverride");
+}
+
+int Screen::widthForBinding(ScriptState* script_state) const {
+  return MaybeIntOverride(script_state, width(), "screen", "widthOverride");
+}
+
 unsigned Screen::colorDepth() const {
   // "If the user agent does not know the color depth or does not want to
   // return it for privacy considerations, it should return 24."
diff --git a/third_party/blink/renderer/core/frame/screen.h b/third_party/blink/renderer/core/frame/screen.h
index 18ecbf9616ad4a356642eacbb06bca3e403e5029..a29b20172c019967de95eaf3a20af7d922bcf4c4 100644
--- a/third_party/blink/renderer/core/frame/screen.h
+++ b/third_party/blink/renderer/core/frame/screen.h
@@ -44,6 +44,7 @@ struct ScreenInfo;
 namespace blink {
 
 class LocalDOMWindow;
+class ScriptState;
 class ScreenScreenOrientation;
 
 class CORE_EXPORT Screen : public EventTarget, public ExecutionContextClient {
@@ -58,6 +59,8 @@ class CORE_EXPORT Screen : public EventTarget, public ExecutionContextClient {
 
   int height() const;
   int width() const;
+  int heightForBinding(ScriptState*) const;
+  int widthForBinding(ScriptState*) const;
   unsigned colorDepth() const;
   unsigned pixelDepth() const;
   int availLeft() const;
diff --git a/third_party/blink/renderer/core/frame/screen.idl b/third_party/blink/renderer/core/frame/screen.idl
index c4c9ee399362c78bfcaf74e87d39e7d02db04c57..3b269e9ac4748d28e76e8fa80d188b640f7134ff 100644
--- a/third_party/blink/renderer/core/frame/screen.idl
+++ b/third_party/blink/renderer/core/frame/screen.idl
@@ -33,8 +33,8 @@
 ] interface Screen : EventTarget {
     [Measure] readonly attribute long availWidth;
     [Measure] readonly attribute long availHeight;
-    [Measure] readonly attribute long width;
-    [Measure] readonly attribute long height;
+    [Measure, CallWith=ScriptState, ImplementedAs=widthForBinding] readonly attribute long width;
+    [Measure, CallWith=ScriptState, ImplementedAs=heightForBinding] readonly attribute long height;
     [Measure] readonly attribute unsigned long colorDepth;
     // pixelDepth() is an alias for colorDepth(), no need to instrument it twice.
     [Measure] readonly attribute unsigned long pixelDepth;
diff --git a/third_party/blink/renderer/core/html/html_element.cc b/third_party/blink/renderer/core/html/html_element.cc
index 93c86594ed368f9055626ff3811c6ec568871659..f5248aaad54e4d8d991613b44efb56a46e8b7cec 100644
--- a/third_party/blink/renderer/core/html/html_element.cc
+++ b/third_party/blink/renderer/core/html/html_element.cc
@@ -32,6 +32,7 @@
 #include "third_party/blink/public/mojom/forms/form_control_type.mojom-blink.h"
 #include "third_party/blink/renderer/bindings/core/v8/js_event_handler_for_content_attribute.h"
 #include "third_party/blink/renderer/bindings/core/v8/v8_binding_for_core.h"
+#include "third_party/blink/renderer/bindings/core/v8/v8_script_runner.h"
 #include "third_party/blink/renderer/bindings/core/v8/v8_show_popover_options.h"
 #include "third_party/blink/renderer/bindings/core/v8/v8_toggle_popover_options.h"
 #include "third_party/blink/renderer/bindings/core/v8/v8_union_boolean_togglepopoveroptions.h"
@@ -125,6 +126,7 @@
 #include "third_party/blink/renderer/core/timing/soft_navigation_heuristics.h"
 #include "third_party/blink/renderer/core/trustedtypes/trusted_script.h"
 #include "third_party/blink/renderer/core/xml_names.h"
+#include "third_party/blink/renderer/platform/bindings/custom_override_helpers.h"
 #include "third_party/blink/renderer/platform/bindings/exception_state.h"
 #include "third_party/blink/renderer/platform/heap/garbage_collected.h"
 #include "third_party/blink/renderer/platform/heap/thread_state.h"
@@ -3419,33 +3421,73 @@ int HTMLElement::OffsetTopOrLeft(bool top) {
   return AdjustedOffsetForZoom(offset);
 }
 
-int HTMLElement::offsetLeftForBinding() {
-  return OffsetTopOrLeft(/*top=*/false);
+namespace {
+int MaybeIntOverride(ScriptState* script_state,
+                     int value,
+                     const char* interface_key,
+                     const char* override_name) {
+  if (!script_state || !script_state->ContextIsValid())
+    return value;
+
+  v8::Isolate* isolate = script_state->GetIsolate();
+  v8::HandleScope handle_scope(isolate);
+  ScriptState::Scope scope(script_state);
+
+  v8::Local<v8::Function> override_fn =
+      GetCustomOverrideFunction(script_state, interface_key, override_name);
+  if (override_fn.IsEmpty())
+    return value;
+
+  v8::Local<v8::Context> context = script_state->GetContext();
+  v8::Local<v8::Value> argv[] = {
+      v8::Number::New(isolate, static_cast<double>(value))};
+  v8::TryCatch try_catch(isolate);
+  v8::Local<v8::Value> override_result;
+  if (!V8ScriptRunner::CallFunction(
+           override_fn, ExecutionContext::From(script_state),
+           context->Global(), 1, argv, isolate)
+           .ToLocal(&override_result))
+    return value;
+
+  if (override_result->IsNumber())
+    return static_cast<int>(override_result.As<v8::Number>()->Value());
+  return value;
+}
+}  // namespace
+
+int HTMLElement::offsetLeftForBinding(ScriptState* script_state) {
+  int value = OffsetTopOrLeft(/*top=*/false);
+  return MaybeIntOverride(script_state, value, "htmlElement",
+                          "offsetLeftOverride");
 }
 
-int HTMLElement::offsetTopForBinding() {
-  return OffsetTopOrLeft(/*top=*/true);
+int HTMLElement::offsetTopForBinding(ScriptState* script_state) {
+  int value = OffsetTopOrLeft(/*top=*/true);
+  return MaybeIntOverride(script_state, value, "htmlElement",
+                          "offsetTopOverride");
 }
 
-int HTMLElement::offsetWidthForBinding() {
+int HTMLElement::offsetWidthForBinding(ScriptState* script_state) {
   GetDocument().EnsurePaintLocationDataValidForNode(
       this, DocumentUpdateReason::kJavaScript);
   int result = 0;
   if (const auto* layout_object = GetLayoutBoxModelObject()) {
     result = AdjustedOffsetForZoom(layout_object->OffsetWidth());
   }
-  return result;
+  return MaybeIntOverride(script_state, result, "htmlElement",
+                          "offsetWidthOverride");
 }
 
 DISABLE_CFI_PERF
-int HTMLElement::offsetHeightForBinding() {
+int HTMLElement::offsetHeightForBinding(ScriptState* script_state) {
   GetDocument().EnsurePaintLocationDataValidForNode(
       this, DocumentUpdateReason::kJavaScript);
   int result = 0;
   if (const auto* layout_object = GetLayoutBoxModelObject()) {
     result = AdjustedOffsetForZoom(layout_object->OffsetHeight());
   }
-  return result;
+  return MaybeIntOverride(script_state, result, "htmlElement",
+                          "offsetHeightOverride");
 }
 
 Element* HTMLElement::unclosedScrollParent() {
diff --git a/third_party/blink/renderer/core/html/html_element.h b/third_party/blink/renderer/core/html/html_element.h
index e9880943520543a402d5487a605cde049fb1d77f..386dff9fb206cb547d3fd2db42ec6e462fab0269 100644
--- a/third_party/blink/renderer/core/html/html_element.h
+++ b/third_party/blink/renderer/core/html/html_element.h
@@ -243,10 +243,10 @@ class CORE_EXPORT HTMLElement : public Element {
   // offsetTopForBinding have their values adjusted for this as well.
   Element* unclosedScrollParent();
   Element* unclosedOffsetParent();
-  int offsetLeftForBinding();
-  int offsetTopForBinding();
-  int offsetWidthForBinding();
-  int offsetHeightForBinding();
+  int offsetLeftForBinding(ScriptState*);
+  int offsetTopForBinding(ScriptState*);
+  int offsetWidthForBinding(ScriptState*);
+  int offsetHeightForBinding(ScriptState*);
 
   ElementInternals* attachInternals(ExceptionState& exception_state);
   virtual FormAssociated* ToFormAssociatedOrNull() { return nullptr; }
diff --git a/third_party/blink/renderer/core/html/html_element.idl b/third_party/blink/renderer/core/html/html_element.idl
index cae1768257bbaf0430362f17211b442020685f26..b60c3beb329f899eceffc2597778ee08405c3b67 100644
--- a/third_party/blink/renderer/core/html/html_element.idl
+++ b/third_party/blink/renderer/core/html/html_element.idl
@@ -62,10 +62,10 @@
     // https://drafts.csswg.org/cssom-view/#extensions-to-the-htmlelement-interface
     [RuntimeEnabled=HTMLElementScrollParent, ImplementedAs=unclosedScrollParent] readonly attribute Element? scrollParent;
     [PerWorldBindings, ImplementedAs=unclosedOffsetParent] readonly attribute Element? offsetParent;
-    [ImplementedAs=offsetTopForBinding] readonly attribute long offsetTop;
-    [ImplementedAs=offsetLeftForBinding] readonly attribute long offsetLeft;
-    [ImplementedAs=offsetWidthForBinding] readonly attribute long offsetWidth;
-    [ImplementedAs=offsetHeightForBinding] readonly attribute long offsetHeight;
+    [CallWith=ScriptState, ImplementedAs=offsetTopForBinding] readonly attribute long offsetTop;
+    [CallWith=ScriptState, ImplementedAs=offsetLeftForBinding] readonly attribute long offsetLeft;
+    [CallWith=ScriptState, ImplementedAs=offsetWidthForBinding] readonly attribute long offsetWidth;
+    [CallWith=ScriptState, ImplementedAs=offsetHeightForBinding] readonly attribute long offsetHeight;
 
     // The Popover API
     [MeasureAs=ElementTogglePopover,RaisesException] boolean togglePopover(optional (TogglePopoverOptions or boolean) options);
